### PR TITLE
Remove `Diagnostic` class.

### LINKED ISSUE #532: Use dict rather than tuple for callbacks?
A full callback set has four functions, `_start`, `_pretask`, `_posttask`, and `_finish`.  We represent them as a tuple with four functions in that order.  

Thoughts on switching to a named entity rather than an ordered one?  Either a dict or namedtuple might be pleasant.  A dict has the advantage that one could more naturally specify a single callback or two.  I currently find myself writing

```python
def post(...)
    ...

cb = [None, None, post, None]
```

When I might prefer to write

```python
cb = {'post': post}
```

Additionally, thoughts on changing the names to `start`, `pre`, `post`, and `finish`?

cc @jcrist

### PR DESCRIPTION
**Pull request by jcrist**
Now everything can be done with single `Callback` class. This also allows creation of callback by passing in functions instead of subclassing:

```python
>>> from dask.callbacks import Callback

>>> def debugit(*args):
...     import pdb; pdb.set_trace()

>>> debug = Callback(posttask=debugit)

>>> with debug:
...     res.compute()
```

Fixes #532.

### CODE REVIEW

File: dask/callbacks.py [hunk 1]
```diff
@@ -1,27 +1,61 @@
...
+
+
+class Callback(object):
+    """Base class for using the callback mechanism."""
+
+    def __init__(self, start=None, pretask=None, posttask=None, finish=None):
+        self._start = start
+        self._pretask = pretask
+        self._posttask = posttask
+        self._finish = finish
```
Comment (mrocklin) [inline dask/callbacks.py]: Should these be public?

File: docs/source/diagnostics.rst [hunk 2]
```diff
@@ -151,3 +151,15 @@ This can now be used as a contextmanager during computation:
...
+
+.. code-block:: python
+
+    >>> def printkeys(key, dask, state):
+    ...    print("Computing: {0}!".format(repr(key)))
+    >>> with Callback(pretask=printkeys):
+    ...     get(dsk, 'c')
+    Computing 'a'!
+    Computing 'b'!
+    Computing 'c'!
```
Comment (mrocklin) [inline docs/source/diagnostics.rst]: Nice

File: dask/diagnostics/core.py [hunk 3]
```diff
@@ -1,26 +0,0 @@
...
-        return self
-
-    def __exit__(self, *args):
-        self._cm.__exit__(*args)
-
-    def register(self):
-        _globals['callbacks'].add(self._callback)
-
-    def unregister(self):
-        _globals['callbacks'].remove(self._callback)
```
Comment (mrocklin) [inline dask/diagnostics/core.py]: Hooray for removing a class!

File: dask/callbacks.py (hunk 1 above)
Comment (jcrist) [inline dask/callbacks.py]: I set them private, as the user should never directly call these. In general, I feel that things like:

```python
In[1]: prof = Profiler()
In[2]: prof.<TAB>
```
should only show things the user should interact with.

**Comment by mrocklin** [comment]
OK, merging soon if no comments

**Pull request merged**